Core Data: Enforce all write operations in the background

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14641 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds the last updates to the storage layer to enforce all write operations in the background:
- Removed `writerDerivedStorage` and `saveDerivedType` to make it official from now on that we have to use `performAndSave` for all write operations.
- Removed `saveIfNeeded` as we no longer need to use this method directly. This comes with a few changes:
  - The method is kept in the extension of `NSManagedObjectContext` to be used by `CoreDataManager`.
  - Updated unit tests for `CoreDataManager`.
  - Since many of our tests still use this method when writing data, I've added an extension method to support this in `MockStorageManager` and marked it as deprecated to avoid fixing all the tests.
  - Updated tests that require the background context for writing to use `saveDerivedType` to avoid the unnecessary warnings about `saveIfNeeded`.
- Enabled the `-enforce-core-data-write-in-background` flag by default to crash the app in debug mode if one attempts to write data in the main thread (view context).
- Updated the behavior of `CoreDataManager`'s `reset` method: 
  - Removed the changes made to the view context as this is unnecessary with our new setup. Now we only need to clear the writer context and save the changes - the view context would then get the updates immediately from the persistent container.
  - Removed the [`reset`](https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext/1506807-reset) call to the background context. This method essentially resets unsaved changes made to a context and would not affect the persistent container. Deleting all stored items should be sufficient for clearing the persistent container.

Additionally, to avoid undefined behavior, the `DefaultStoresManager`'s `deauthenticate` method was updated to reset the session manager first before setting the state to deauthenticated. The wrong order causes a case where the user is unauthenticated with a selected store, which [triggers](https://github.com/woocommerce/woocommerce-ios/blob/f30517c85795ad30e860e12266300ca7ed202b13/WooCommerce/Classes/ViewRelated/AppCoordinator.swift#L85) `deauthenticate` one more time.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Build and run the app.
- Do some quick smoke tests to confirm that everything works as expected.
- Log out of the app to confirm that it works correctly.
- Log back in to confirm that data is freshly fetched again.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
Tested on simulator iPhone 16 Pro iOS 18.1 and confirmed that the app works correctly when logging in/out and loading data.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Update MockOrderActionHandler and MockProductActionHandler to write data using performAndSave

itsmeichigo committed:
Remove writerDerivedStorage and saveDerivedType from StorageManagerType interface

itsmeichigo committed:
Rename writerDerivedStorage in CoreDataManager and make it private

itsmeichigo committed:
Remove saveDerivedType from CoreDataManager

itsmeichigo committed:
Remove deprecated reference to writerDerivedStorage in GoogleAdsStore, ProductVariationStore, and ReceiptStore

itsmeichigo committed:
Fix test build failures due to reference of writerDerivedStorage

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Remove redundant save for view storage when resetting the storage

itsmeichigo committed:
Enable -enforce-core-data-write-in-background by default for debug builds

itsmeichigo committed:
Update CoreDataManagerTests to use performAndSave instead of writing to view storage

itsmeichigo committed:
Remove redundant save in MockProductReviewAction

itsmeichigo committed:
Remove saveIfNeeded in StorageType interface

itsmeichigo committed:
Update tests that use the background context

itsmeichigo commented on Storage/Storage/CoreData/CoreDataManager.swift:
```diff
@@ -116,12 +107,12 @@ public final class CoreDataManager: StorageManagerType {
                 operation(derivedStorage)
 
-                derivedStorage.saveIfNeeded()
+                (derivedStorage as! NSManagedObjectContext).saveIfNeeded()
```
Intentionally using forced unwrapping here as we don't expect the writer storage to be of any other type in `CoreDataManager`.

itsmeichigo commented on Storage/Storage/CoreData/CoreDataManager.swift:
```diff
@@ -163,7 +154,6 @@ public final class CoreDataManager: StorageManagerType {
         viewContext.performAndWait {
             viewContext.reset()
             self.deleteAllStoredObjects(in: viewContext)
-            viewContext.saveIfNeeded()
```
This is not needed as the view storage should only get merges from the persistent container automatically.

itsmeichigo committed:
Update reset method for CoreDataManager to not reset changes

itsmeichigo committed:
Deauthenticate: reset the session manager first before updating state to avoid triggering deauthenticate twice

In AppCoordinate, we're re-deauthenticate the user when the user is deauthenticated with a selected store. Reordering the steps when deauthenticating can avoid this.

itsmeichigo committed:
Update comments where the authentication state is checked before reloading data

itsmeichigo committed:
Update release notes

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:
```diff
@@ -532,6 +532,13 @@ private extension DashboardViewModel {
 
     func configureOrdersResultController() {
         func refreshHasOrders() {
+            /// Upon logging out, `CoreDataManager` clears the storage triggering data change.
+            /// Checking the authentication state helps avoiding reloading data
+            /// in the unauthenticated state.
+            guard stores.isAuthenticated else {
+                return
+            }
```
Moved this here as previously we were checking only when the storage is reset. Now this is needed when the storage is cleared.

itsmeichigo commented on WooCommerce/Classes/Blaze/BlazeLocalNotificationScheduler.swift:
```diff
@@ -177,6 +171,12 @@ private extension DefaultBlazeLocalNotificationScheduler {
     }
 
     func scheduleNoCampaignReminderIfNeeded() {
+        /// Upon logging out, `CoreDataManager` clears the storage triggering data change.
+        /// Checking the authentication state helps avoiding reloading data
+        /// in the unauthenticated state.
+        guard stores.isAuthenticated else {
+            return
+        }
```
Moved this here as previously we were checking only when the storage is reset. Now this is needed when the storage is cleared.

itsmeichigo committed:
Update doc for the storage layer with the default launch arguments

selanthiraiyan commented on docs/STORAGE.md:
```diff
@@ -25,6 +25,8 @@ Notes:
 - For thread safety, do not send any `NSManagedObject` instance to the completion closure of `performAndSave`. There's an assertion to ensure at debug runtime this does not happen.
 - For performance reasons, please be mindful with fetch requests. Avoid making multiple fetch requests in for loops. This can be replaced by a single fetch request for a list of objects instead.
 - For attributes of type Transformable, if the transformer is `NSSecureUnarchiveFromData`, make sure to input a `class` type in the Custom Class field in the Core Data model. Using a Swift type (e.g. [Int64] or [String]) would cause an error logged in Xcode 16.
+- We have a launch argument `-enforce-core-data-write-in-background` enabled by default in the WooCommerce scheme. This will intentionally crashes the app in the debug mode when one attempts to perform write operations in the main thread with the view context.
```
```suggestion
- We have a launch argument `-enforce-core-data-write-in-background` enabled by default in the WooCommerce scheme. This will intentionally crash the app in the debug mode when one attempts to perform write operations in the main thread with the view context.
```

selanthiraiyan commented on Storage/Storage/CoreData/CoreDataManager.swift:
```diff
@@ -87,23 +87,14 @@ public final class CoreDataManager: StorageManagerType {
     public let viewStorage: StorageType
 
-    /// Returns a shared derived storage instance dedicated for write operations.
+    /// Returns a storage instance dedicated for write operations.
```
Opinion: We can remove `Returns` as this is no longer a `public` property.

itsmeichigo replied:
Updated in 61c8e20a6a50027a52744ac976edc415c4f15238

selanthiraiyan commented on Storage/Storage/CoreData/CoreDataManager.swift:
```diff
@@ -142,42 +133,34 @@ public final class CoreDataManager: StorageManagerType {
+    /// This method effectively **deletes** all of the stored data from the persistent container,
+    /// and generates a blank Persistent Store from scratch.
+    /// We don't use `NSManagedObjectContext.reset` here as it does not affect the persistent container,
+    /// only reverts the changes made on a particular context.
```
I think we don't have to explain this as we no longer use the `NSManagedObjectContext.reset` method. WDYT?

itsmeichigo replied:
Also updated in 61c8e20a6a50027a52744ac976edc415c4f15238

selanthiraiyan commented on Storage/Storage/CoreData/CoreDataManager.swift:
```diff
@@ -142,42 +133,34 @@ public final class CoreDataManager: StorageManagerType {
-            backgroundContext.reset()
         }, completion: {
             DDLogVerbose("💣 [CoreDataManager] Stack Destroyed!")
-            NotificationCenter.default.post(name: .StorageManagerDidResetStorage, object: self)
```
Could you help me understand why we don't need this notification anymore?

itsmeichigo replied:
This notification is relevant only if we reset the storage. In this case we removed the `reset` calls so I don't think it's necessary.

selanthiraiyan commented on Yosemite/YosemiteTests/Mocks/MockStorageManager.swift:
```diff
@@ -56,7 +56,7 @@ public class MockStorageManager: StorageManagerType {
     /// This method effectively destroys all of the stored data, and generates a blank Persistent Store from scratch.
     ///
-    public func reset() {
+    public func reset(onCompletion: (() -> Void)?) {
```
I think we should also fire the `onCompletion` block from this method.

itsmeichigo replied:
Done in e06168d3f9c50711fe4f0ca8260a50a48deebd14.

selanthiraiyan reviewed (approved):
Thanks for working on this, Huong! The changes look good to me and tests well. I left a few questions and non-blocking comments.

itsmeichigo committed:
Fix typo in docs

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

itsmeichigo committed:
Update comments in CoreDataManager

itsmeichigo committed:
Trigger onCompletion for MockStorageManager's reset
